### ISSUE TITLE
S3 Presigned PutObjectCommand URLs ignore Sha256 Hash when uploading

### ISSUE THREAD

**Issue by samjarman** [aws/aws-sdk-js-v3#4448]
### Checkboxes for prior research

- [X] I've gone through [Developer Guide](https://docs.aws.amazon.com/sdk-for-javascript/v3/developer-guide) and [API reference](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest)
- [X] I've checked [AWS Forums](https://forums.aws.amazon.com) and [StackOverflow](https://stackoverflow.com/questions/tagged/aws-sdk-js).
- [X] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-js-v3/issues) and didn't find any solution.

### Describe the bug

Up until around January 2023, we were successfully using  the S3 SDK and S3 Request pre-signer. However, it just stopped working, with no code changes and no dependency changes.

This is the code we are using:

```ts
import { ChecksumAlgorithm, PutObjectCommand, PutObjectCommandInput, S3Client } from '@aws-sdk/client-s3';
import { fromIni } from '@aws-sdk/credential-providers';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';

const getPresignedURL = async (bucket: string, key: string, sha256Hash: string): Promise<string> => {
	const s3 = new S3Client({ region: 'eu-west-1', credentials: fromIni({ profile: 'my-profile' }) });

	const params: PutObjectCommandInput = {
		Bucket: bucket,
		Key: key,
		ChecksumSHA256: sha256Hash,
		ChecksumAlgorithm: ChecksumAlgorithm.SHA256,
	};
	const getPutPresignedUrl = new PutObjectCommand(params);
	return getSignedUrl(s3, getPutPresignedUrl, { expiresIn: 300 });
};

const go = async (): Promise<void> => {	
	const url = await getPresignedURL('some-bucket', 'test-file-name.txt', 'not-a-hash'); 

	console.log(url);
};
void go();
```

S3 Support has not been helpful in this matter unfortunately. They referred us to a python blog post, which uses a different technique, more similar to using `@aws-sdk/s3-presigned-post. However, we're happy with and continue to be interested in using PUT over POST for our use case.

### SDK version number

@aws-sdk/s3-request-presigner@3.272.0, @aws-sdk/client-s3@3.272.0

### Which JavaScript Runtime is this issue in?

Node.js

### Details of the browser/Node.js/ReactNative version

v18.13.0

### Reproduction Steps

Run the above code, get a URL.

Paste into POSTMAN, attach a/any file. 

### Observed Behavior

Actual: A 200, file was uploaded to bucket

### Expected Behavior

Expect: To see a 4XX -  Referring to mismatch file integrity

**Comment by yenfryherrerafeliz** [comment]
Hi @samjarman, thanks for opening this issue. Are you providing the hash in the headers when executing the request?. In your case since you are using postman, are you including the header "x-amz-checksum-sha256" with the hash as value when sending the request?

Thanks!

**Comment by samjarman** [comment]
No, it's appended to the url as a query parameter. Has something changed there? Do we need to put the same info in the headers now too?

**Comment by RanVaknin** [comment]
Hi @samjarman,

Presigned urls do not have headers associated with them (they are a url string),  this means that in order to get signed headers, they are hoisted into the query string. 

The SDKs got a shared feature request to implement flexible checksums for s3 uploads. Unfortunately that cross SDK feature request did not have specifications for presigned URLs so there is no support server-side for reading checksum headers from the query string.
We have informed the s3 team about this a while ago and it seems like it only recently got prioritized. 

Right now there is no way to provide a checksum with a presigned url. The service will return a 200 but will silently ignore the checksum validation.

[Related](https://github.com/aws/aws-sdk/issues/338)

I'll see if I can find out more internally.

Thanks,
Ran~
